### PR TITLE
Add `min_new_tokens` argument in generate() implementation

### LINKED ISSUE #20756: min_new_tokens option in generate() implementation
### Feature request

Similarly to the `max_new_tokens`, a `min_new_tokens` option would count only the newly generated tokens, ignoring the tokens of the input sequence (prompt) in decoder only models.

### Motivation

The option `min_length` of the `generate()` method might be ambiguous for decoder only models. It is not clear if decoder only models consider the length of the input (prompt) for the `min_length` condition or only the newly generated tokens.
In Encoder Decoder (seq2seq) it is clear though.

### Your contribution

Not that I remember. But I could test it.

### PR DESCRIPTION
**Pull request by silverriver**
# What does this PR do?

Fixes #20756 #20814 #20614 (cc @gonced8 @kotikkonstantin)

As many said, it is better to add an argument `min_new_tokens` to the `.generate()` method to limit the length of newly generated tokens. The current parameter `min_length` limits the length of `prompt + newly generated tokens`, not the length of `newly generated tokens`.

It seems that all tests are passed

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- @gante

### CODE REVIEW

**Comment by gante** [comment]
(Note: this PR depends on the resolution of https://github.com/huggingface/transformers/issues/20814, so I'm waiting for it before I review this one)

**Comment by silverriver** [comment]
> (Note: this PR depends on the resolution of #20814, so I'm waiting for it before I review this one)

Great, let me know if there are anything I can help.

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1234,6 +1234,17 @@ def generate(
                 "(https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)"
             )
 
+        has_default_min_length = kwargs.get("min_length") is None and generation_config.min_length == 0
```
Comment (gante) [inline src/transformers/generation/utils.py:1237]:
```suggestion
        has_default_min_length = kwargs.get("min_length") is None and generation_config.min_length is not None
```
(some models can have a default value for min_length that is not 0)

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1234,6 +1234,17 @@ def generate(
...
+        if has_default_max_length and generation_config.min_new_tokens is not None:
+            generation_config.min_length = generation_config.min_new_tokens + input_ids_seq_length
```
Comment (gante) [inline src/transformers/generation/utils.py:1239]: We don't need to set `min_length` as we have a dedicated logits processor :)

These lines can be removed and, instead, while setting the logits processor list, we can add a `MinNewTokensLengthLogitsProcessor` as defined in #20892 (e.g. after [this line](https://github.com/huggingface/transformers/blob/305f41e4dea99d9fdc212287263b3748aacc63d3/src/transformers/generation/utils.py#L811))

**Review by gante — COMMENTED** [review]
Added some comments to integrate this changes with the changes from #20892 :)

They are small changes, and we should be able to merge afterward!

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (silverriver) [reply src/transformers/generation/utils.py:1239]: Sure, I will make a commit

**Comment by silverriver** [comment]
Sorry that I have accidently rebased my PR to include the change of #20892. It seems that my operation will link this PR to all pervious commits that have been merged to main.

I think it would be more convenient to make a new PR from the main branch. So I am closing this one.

**Comment by silverriver** [comment]
I have made a new PR #21044
